### DISCUSSION TITLE
Add UI tests?

### DISCUSSION THREAD

**Discussion by nbaars** [WebGoat/WebGoat#1162] [Ideas]
In the past we had specific UI test through a platform which tested all the browsers and platforms.
The problem was that we could not run the tests locally (well not that easy) and the tests were very flaky.

Since then there are new solutions available:

- [Cypress](https://www.cypress.io/)
- [TestCafe](https://testcafe.io/)
- [Playwright](https://playwright.dev/)
- ...

Playwright looks like a good candidate as we can write test in Java. I think the tests should focus solving the assignment is the UI correctly updated are the hints presented etc.

If someone has experience in this area please share your them here.

**Comment by zubcevic** [comment]
Hi @nbaars , @aolle , and others 
Robot framework could also be an option. It is based on python and can be used for all kinds of tests. With the selenium library it can also do UI tests. I have set up a local branch with a robot framework test and tried to get a test working using GitHub Actions.
It is now kind of working: https://github.com/zubcevic/WebGoat/actions/runs/2717924132 
Here is a readme: https://github.com/zubcevic/WebGoat/blob/robot-framework/robot/README.md
And here is the test: https://github.com/zubcevic/WebGoat/blob/robot-framework/robot/goat.robot 

I would suggest to only test the UI elements themselves and not the assignments, otherwise people will start using this in classrooms or competition.

**Reply by zubcevic** [reply]
The workflow is placed here for now, so it acts on my local branch: https://github.com/zubcevic/WebGoat/blob/robot-framework/.github/workflows/build.yml

**Reply by aolle** [reply]
> Hi @nbaars , @aolle , and others Robot framework could also be an option. It is based on python and can be used for all kinds of tests. With the selenium library it can also do UI tests. I have set up a local branch with a robot framework test and tried to get a test working using GitHub Actions. It is now kind of working: https://github.com/zubcevic/WebGoat/actions/runs/2717924132 Here is a readme: https://github.com/zubcevic/WebGoat/blob/robot-framework/robot/README.md And here is the test: https://github.com/zubcevic/WebGoat/blob/robot-framework/robot/goat.robot

It looks good, let's see what Nanne thinks.

> I would suggest to only test the UI elements themselves and not the assignments, otherwise people will start using this in classrooms or competition.

I agree. WebGoat is made for learn, cheating is fooling oneself, definitively you can pass all assignments checking the assignments unit tests actually.

**Reply by nbaars** [reply]
Looks good, like the scripting language.

> I would suggest to only test the UI elements themselves and not the assignments, otherwise people will start using this in classrooms or competition.

Agreed, if we focus on the UI interaction the script will be simple. It will give us the missing UI interaction tests we miss at the moment. I would say let's integrate this branch into the develop branch 👍

**Reply by zubcevic** [reply]
ok I will clean up and squash some of the commits and find a place in the workflow where we can enable this. I added a second browser in the test for web wolf. This can also be done using robot framework

**Reply by zubcevic** [reply]
The syntax is fairly simple. Just get used to having spaces after keywords. And for complex situations a bit of xpath knowledge comes in handy, but that is in most of these UI tools
